build_compilers/lib/src/dev_compiler_bootstrap_builder.dart

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:analyzer/analyzer.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as _p; // ignore: library_prefixes
+
+import 'dev_compiler_builder.dart';
+import 'module_builder.dart';
+import 'modules.dart';
+
+/// Alias `_p.url` to `p`.
+_p.Context get p => _p.url;
+
+final String bootstrapJsExtension = '.dart.bootstrap.js';
+final String jsEntrypointExtension = '.dart.js';
+final String jsEntrypointSourceMapExtension = '.dart.js.map';
+
+class DevCompilerBootstrapBuilder extends Builder {
+  @override
+  final buildExtensions = {
+    '.dart': [
+      bootstrapJsExtension,
+      jsEntrypointExtension,
+      jsEntrypointSourceMapExtension
+    ],
+  };
+
+  @override
+  Future build(BuildStep buildStep) async {
+    var dartEntrypointId = buildStep.inputId;
+    var isAppEntrypoint = await _isAppEntryPoint(dartEntrypointId, buildStep);
+    if (!isAppEntrypoint) return;
+
+    var moduleId = buildStep.inputId.changeExtension(moduleExtension);
+    var module = new Module.fromJson(
+        JSON.decode(await buildStep.readAsString(moduleId))
+            as Map<String, dynamic>);
+
+    // First, ensure all transitive modules are built.
+    await _ensureTransitiveModules(module, buildStep);
+
+    var appModuleName = p.withoutExtension(module.jsId.path);
+
+    // The name of the entrypoint dart library within the entrypoint JS module.
+    //
+    // This is used to invoke `main()` from within the bootstrap script.
+    //
+    // TODO(jakemac53): Sane module name creation, this only works in the most
+    // basic of cases.
+    //
+    // See https://github.com/dart-lang/sdk/issues/27262 for the root issue
+    // which will allow us to not rely on the naming schemes that dartdevc uses
```

Comment:
> 80 chars here and in a few other comments.

Reply:
this one is actually good but I fixed a different one